src/ide/navigation.ts

```diff
@@ -0,0 +1,116 @@
+export interface IPageController {
+	navigate?(newLocation: URL): boolean;
+	checkUnsavedChanges?(): boolean;
+}
+
+export type PageControllerConstructor = {
+	new (container: HTMLElement, location: URL): IPageController;
+};
+
+export type PageResolver = (url: URL) => PageControllerConstructor | null;
+
+let currentController: IPageController | undefined;
+let pageResolver: PageResolver | undefined;
+
+/**
+ * Called on load to show the desired page
+ *
+ * @param url the current URL
+ * @param container the container element to place the page in
+ * @param resolver a function mapping URLs to their controller classes
+ */
+export function initializeNavigation(
+	url: URL,
+	container: HTMLElement,
+	resolver: PageResolver
+): void {
+	if (pageResolver) return;
+
+	const controllerType = resolver(url);
+	if (!controllerType) return;
+
+	currentController = new controllerType(container, url);
+	pageResolver = resolver;
+
+	window.addEventListener("popstate", () => {
+		if (!tryHandleNavigation(new URL(location.href))) location.reload();
+	});
+
+	document.addEventListener("click", (event) => {
+		if (!(event.target instanceof HTMLAnchorElement)) return;
+
+		if (navigate(event.target.href, false)) event.preventDefault();
+	});
+
+	if (currentController.checkUnsavedChanges) {
+		window.addEventListener("beforeunload", (event) => {
+			if (currentController?.checkUnsavedChanges?.()) {
+				event.preventDefault();
+				event.returnValue = "";
+			}
+		});
+	}
+}
+
+function tryHandleNavigation(newUrl: URL): boolean {
```

Comment:
Could you document the idea behind this function? It becomes clear only when checking the navigate function.